File: LrcSeeker/ChineseConverter.m
//
//  ChineseConverter.m
//  LrcSeeker
//
//  Created by Eru on 15/11/21.
//  Copyright © 2015年 Eru. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "ChineseConverter.h"

NSString* convertToChineseUsingProfile (NSString* profile, NSString* inputStr) {
    opencc_t cc = opencc_open(profile.UTF8String);
    const char* cInput = inputStr.UTF8String;
    char *cOutput = opencc_convert_utf8(cc, cInput, strlen(cInput));
    NSString *outputStr = [NSString stringWithUTF8String:cOutput];
    opencc_convert_utf8_free(cOutput);
    opencc_close(cc);
    return outputStr;
}

NSString* convertToSC (NSString* input) {
    NSString *profilePath=[[NSBundle mainBundle] pathForResource:@"t2s" ofType:@"json"];
    return convertToChineseUsingProfile(profilePath, input);
}

NSString* convertToTC (NSString* input) {
    NSString *profilePath=[[NSBundle mainBundle] pathForResource:@"s2t" ofType:@"json"];
    return convertToChineseUsingProfile(profilePath, input);
}

NSString* convertToTCTW (NSString* input) {
    NSString *profilePath=[[NSBundle mainBundle] pathForResource:@"s2tw" ofType:@"json"];
    return convertToChineseUsingProfile(profilePath, input);
}

NSString* convertToTCHK (NSString* input) {
    NSString *profilePath=[[NSBundle mainBundle] pathForResource:@"s2hk" ofType:@"json"];
    return convertToChineseUsingProfile(profilePath, input);
}

File: Lyrics/ChineseConverter.swift
//
//  ChineseConverter.swift
//  Lyrics
//
//  Created by Eru on 15/11/10.
//  Copyright © 2015年 Eru. All rights reserved.
//

import Cocoa

func convertToSC(_ input: String) -> String {
    let profilePath: NSString = Bundle.main.path(forResource: "t2s", ofType: "json")! as NSString
    return convertToChineseUsingProfile(profilePath,inputStr: input as NSString)
}

func convertToTC(_ input: String) -> String {
    let profilePath: NSString = Bundle.main.path(forResource: "s2t", ofType: "json")! as NSString
    return convertToChineseUsingProfile(profilePath,inputStr: input as NSString)
}

func convertToTC_TW(_ input: String) -> String {
    let profilePath: NSString = Bundle.main.path(forResource: "s2tw", ofType: "json")! as NSString
    return convertToChineseUsingProfile(profilePath,inputStr: input as NSString)
}

func convertToTC_HK(_ input: String) -> String {
    let profilePath: NSString = Bundle.main.path(forResource: "s2hk", ofType: "json")! as NSString
    return convertToChineseUsingProfile(profilePath,inputStr: input as NSString)
}

private func convertToChineseUsingProfile(_ profile: NSString, inputStr: NSString) -> String {
    let cc: opencc_t = opencc_open(profile.utf8String)
    let cInput: UnsafePointer<Int8> = inputStr.utf8String!
    let cOutputStr: UnsafeMutablePointer<Int8> = opencc_convert_utf8(cc, cInput, Int(strlen(cInput)))
    let outputStr: String? = String(cString: cOutputStr)
    opencc_convert_utf8_free(cOutputStr)
    opencc_close(cc)
    if outputStr == nil {
        return ""
    } else {
        return outputStr!
    }
}
